test/e2e/custom_wait_rules_test.go

```diff
@@ -0,0 +1,129 @@
+      funcContractV1:
+        resource.star: |
+          def is_done(resource):
+              state = resource.status.currentState
+              if state == "Failed":
+                return {"done": True, "successful": False, "message": ""}
+              elif state == "Running":
+                return {"done": True, "successful": True, "message": ""}
+              else:
+                return {"done": True, "successful": False, "message": "Not in Failed or Running state"}
+              end
+          end
+    resourceMatchers:
+      - apiVersionKindMatcher: {apiVersion: stable.example.com/v1, kind: CronTab}
+`
+
+	crdYaml := `
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: crontabs.stable.example.com
+spec:
+  group: stable.example.com
+  versions:
+    - name: v1
+      served: true
+      storage: true
+      schema:
+        openAPIV3Schema:
+          type: object
+          properties:
+            spec:
+              type: object
+              properties:
+                cronSpec:
+                  type: string
+                image:
+                  type: string
+                replicas:
+                  type: integer
+            status:
+              type: object
+              properties:
+                currentState:
+                  type: string
+  scope: Namespaced
+  names:
+    plural: crontabs
+    singular: crontab
+    kind: CronTab
+---
+`
+	crYaml := `
+apiVersion: "stable.example.com/v1"
+kind: CronTab
+metadata:
+  name: my-new-cron-object-1
+spec:
+  cronSpec: "* * * * */5"
+  image: my-awesome-cron-image
+status:
+  currentState: %s
+---
+`
+
+	name := "test-custom-wait-rule-contract-v1"
+	cleanUp := func() {
+		kapp.Run([]string{"delete", "-a", name})
+	}
+
+	cleanUp()
+	defer cleanUp()
+
+	logger.Section("deploy resource with current state as running", func() {
+		_, err := kapp.RunWithOpts([]string{"deploy", "-f", "-", "-a", name}, RunOpts{
+			StdinReader: strings.NewReader(crdYaml + fmt.Sprintf(crYaml, "Running") + config)})
+		if err != nil {
+			require.Errorf(t, err, "Expected CronTab to be deployed")
+		}
+		NewPresentClusterResource("CronTab", "my-new-cron-object-1", env.Namespace, kubectl)
+	})
+
+	cleanUp()
+
+	logger.Section("deploy resource with current state as failed", func() {
+		_, err := kapp.RunWithOpts([]string{"deploy", "-f", "-", "-a", name}, RunOpts{
+			StdinReader: strings.NewReader(crdYaml + fmt.Sprintf(crYaml, "Failed") + config),
+			AllowError:  true,
+		})
+
+		require.Contains(t, err.Error(), "kapp: Error: waiting on reconcile crontab/my-new-cron-object-1")
+	})
+
+	cleanUp()
+
+	logger.Section("deploy resource with current state as invalid", func() {
+		_, err := kapp.RunWithOpts([]string{"deploy", "-f", "-", "-a", name}, RunOpts{
+			StdinReader: strings.NewReader(crdYaml + fmt.Sprintf(crYaml, "Invalid") + config),
+			AllowError:  true,
+		})
+
+		require.Contains(t, err.Error(), "Not in Failed or Running state")
+	})
```

Comment:
instead of a separate section why dont we change wait rule to include message when it's failure? actually we should probably do same for success and check on the message as well.

Reply:
Done, Made the changes to check for message also both in case of success and failure.